src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs

```diff
@@ -480,6 +480,83 @@ public void TestNameIntrospectsArrayValues()
             });
         }
 
+        [TestCaseSource(nameof(ExplicitTypeArgsWithUnrelatedParametersTestCases))]
+        public void ExplicitTypeArgsWithUnrelatedParameters<T>(string input)
+        {
+            Assert.That(typeof(T), Is.EqualTo(typeof(long)));
+            Assert.That(input, Is.EqualTo("2"));
+        }
+
+        private static IEnumerable<TestCaseData> ExplicitTypeArgsWithUnrelatedParametersTestCases()
+        {
+            yield return new TestCaseData("2") { TypeArgs = new[] { typeof(long) } };
+        }
+
+        [TestCaseSource(nameof(GenericMethodAndParameterWithExplicitOrImplicitTypingTestCases))]
+        public Type GenericMethodAndParameterWithExplicitOrImplicitTyping<T>(T input)
+            => typeof(T);
+
+        private static IEnumerable<TestCaseData> GenericMethodAndParameterWithExplicitOrImplicitTypingTestCases()
+        {
+            yield return new TestCaseData(2)
+            {
+                TypeArgs = new[] { typeof(long) },
+                ExpectedResult = typeof(long)
+            };
+            yield return new TestCaseData(2L)
+            {
+                TypeArgs = new[] { typeof(long) },
+                ExpectedResult = typeof(long)
+            };
+            yield return new TestCaseData(2)
+            {
+                ExpectedResult = typeof(int)
+            };
+            yield return new TestCaseData(2L)
+            {
+                ExpectedResult = typeof(long)
+            };
+        }
+
+        [Test]
+        public void ExplicitTypeArgsWithUnassignableParametersFailsAtRuntime()
+        {
+            var suite = TestBuilder.MakeParameterizedMethodSuite(
+                typeof(TestCaseSourceAttributeFixture),
+                nameof(TestCaseSourceAttributeFixture.MethodWithIncompatibleGenericTypeAndArgument));
+
+            var test = suite.Tests[0];
+
+            Assert.That(test.RunState, Is.EqualTo(RunState.Runnable));
+            var exception = Assert.Throws<NUnitException>(() => test.Method!.Invoke(test.Parent, test.Arguments));
+
+            Assert.That(exception.InnerException, Has.Message.EqualTo("Object does not match target type."));
```

Comment:
This is wrong. The test now fails because `test.Parent` is of type `ParameterizedMethodSuite` and not `TestCaseSourceAttributeFixture`
The error talks about `target` type which is the `this` argument, not about parameters.
You have to call `TestBuilder.RunTest` and check the results which shows that the test fails.
However, in case of mismatch of type arguments, the test is marked as non-runnable as you don't have to run it at al.

Reply:
Sorry, maybe there's been a misunderstanding on my end. I like your suggestion as it avoids the need to explicitly check exception status. I'll push some changes shortly.

> However, in case of mismatch of type arguments, the test is marked as non-runnable as you don't have to run it at al.

I may still not be understanding this suggestion. Are you saying that mismatched values and generic types should fail earlier than test execution time? If so, I think I'm having a hard time reconciling that with "let the (dotnet) runtime sort it out" when `TypeArgs` are explicitly passed.

Reply:
> Are you saying that mismatched values and generic types should fail earlier than test execution time

Yes. If NUnit cannot determine the TypeArgs. See [NUnitTestCaseBuilder](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/Builders/NUnitTestCaseBuilder.cs#L222-L223). If NUnit cannot determine the TypeArguments to use, the test is not runnable.

> "let the (dotnet) runtime sort it out" when TypeArgs are explicitly passed.

If `TypeArgs` is passed, the above `TryGetTypeArguments` is not called and NUnit first try to do some conversion in `TypeHelper.ConvertArgumentList` but after that passes it on to the runtime. The conversions done by NUnit are the ones normally done by the compiler when passing an `int` to a `double` parameter.